packages/@aws-cdk/aws-iam/lib/policy-document.ts

```diff
@@ -197,12 +197,20 @@ export class PolicyStatement extends Token {
    * Indicates if this permission has a "Principal" section.
    */
   public get hasPrincipal() {
-    return this.principal && this.principal.length > 0;
+    return Object.keys(this.principal).length > 0;
   }
 
   public addPrincipal(principal: PolicyPrincipal): PolicyStatement {
     const fragment = principal.policyFragment();
-    this.principal.push(fragment.principalJson);
+    for (const key of Object.keys(fragment.principalJson)) {
+      this.principal[key] = this.principal[key] || [];
+      const value = fragment.principalJson[key];
+      if (Array.isArray(value)) {
+        this.principal[key].push(...value);
```

Comment:
What guarantees that `this.principal[key]` is an array? What if it is `{ AWS: "*" }` ?

Reply:
The only way you can set `this.principal[key]` is though this method, which only ever sets it to an array. It also happens to be typed like that, so `tsc` kind of guarantees this for me.

Arrays of a single value would be replaced by just that value at `toJSON` time, same as `{ AWS:  '*' }` is replaced by `*` at `toJSON` time.